### PR TITLE
Connection leak on failure to parse body

### PR DESCRIPTION
**Pull request by lukasniemeier-zalando**
In case the DefaultMessageReader fails to parse the response body, connections are not returned to the connection pool and stay leased forever.

## Description

A test has been added to show the following scenario:

* the `DefaultMessageReader` fails to parse the response body
* a `RestClientException is raised`
* client connection is not returned to the connection pool and stay leased forever
* connection pool gets exhausted

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: riptide-core/src/main/java/org/zalando/riptide/DefaultMessageReader.java [hunk 1]
```diff
@@ -44,9 +44,13 @@
     private <I> I readBody(final Type type, final ClientHttpResponse response) throws IOException {
         final ResponseExtractor<I> extractor = new HttpMessageConverterExtractor<>(type, converters);
         try {
-            return extractor.extractData(response);
+            try {
+                return extractor.extractData(response);
+            } catch (final IOException | RuntimeException e) {
+                response.close();
```
Comment (whiskeysierra) [inline riptide-core/src/main/java/org/zalando/riptide/DefaultMessageReader.java]: If this fails with its own `IOException` we might loose the original one. Would that be an issue or can we ignore that?

File: riptide-httpclient/src/test/java/org/zalando/riptide/httpclient/AbstractApacheClientHttpRequestFactoryTest.java [hunk 2]
```diff
@@ -179,6 +186,27 @@ void shouldReleaseConnection() {
         });
     }
 
+    @Test
+    void shouldReleaseConnectionOnFailureToReadBody() {
+        driver.addExpectation(onRequestTo("/wrong-content-type"), giveResponse("[]", "text/plain"));
+        driver.addExpectation(onRequestTo("/wrong-content-type"), giveResponse("[]", "text/plain"));
+
+        final Runnable request = throwingRunnable(() -> {
```
Comment (whiskeysierra) [inline riptide-httpclient/src/test/java/org/zalando/riptide/httpclient/AbstractApacheClientHttpRequestFactoryTest.java]:
```suggestion
        final ThrowingRunnable<Exception> request = () -> {
```

Comment (lukasniemeier-zalando) [reply riptide-httpclient/src/test/java/org/zalando/riptide/httpclient/AbstractApacheClientHttpRequestFactoryTest.java]: On the one hand "why not" but on the other hand "why"?

File: riptide-core/src/main/java/org/zalando/riptide/DefaultMessageReader.java (hunk 1 above)
Comment (lukasniemeier-zalando) [reply riptide-core/src/main/java/org/zalando/riptide/DefaultMessageReader.java]: Aha! Another reason `SneakyThrows` must go: I've checked the contract of `ClientHttpResponse` which specifies that implementations must swallow `IOExceptions`, hence I went without another `try-catch` and possibly a "suppressed exception". Enter Riptide's `ApacheClientHttpResponse` which actually throws `IOExceptions` :disappointed: 

IMO we should rather change `ApacheClientHttpResponse` to not throw `IOExceptions`. 

You can still have a case for any `RuntimeException` obviously, but in that case I favor swallowing the original one as this is an unexpected scenario.

File: riptide-httpclient/src/test/java/org/zalando/riptide/httpclient/AbstractApacheClientHttpRequestFactoryTest.java (hunk 2 above)
Comment (whiskeysierra) [reply riptide-httpclient/src/test/java/org/zalando/riptide/httpclient/AbstractApacheClientHttpRequestFactoryTest.java]: I usually prefer to use the type over the functions. But it's actually shorter...

File: riptide-core/src/main/java/org/zalando/riptide/DefaultMessageReader.java (hunk 1 above)
Comment (whiskeysierra) [reply riptide-core/src/main/java/org/zalando/riptide/DefaultMessageReader.java]: > IMO we should rather change `ApacheClientHttpResponse` to not throw `IOExceptions`.

Sounds good

File: riptide-core/src/main/java/org/zalando/riptide/DefaultMessageReader.java [hunk 3]
```diff
@@ -44,7 +44,12 @@
...
+                throw e;
+            }
         } catch (final RestClientException e) {
             propagateIfPossible(e.getCause(), IOException.class, HttpMessageNotReadableException.class);
```
Comment (whiskeysierra) [inline riptide-core/src/main/java/org/zalando/riptide/DefaultMessageReader.java]: Now that we figured out why we have this statement, maybe we should put a comment on it. At least for our future selves.

**Comment by whiskeysierra** [comment]
:+1:

**Pull request merged**